YANG Doctor Review

haomianzheng opened the pull request:
this pull request is to fix the comments proposed in https://mailarchive.ietf.org/arch/msg/ccamp/vpIjwy2HokZI74pKkIBrr0fziIw, and its followup discussion.

haomianzheng committed:
YANG Doctor Review

this pull request is to fix the comments proposed in https://mailarchive.ietf.org/arch/msg/ccamp/vpIjwy2HokZI74pKkIBrr0fziIw, and its followup discussion

agva123 reviewed (commented):
layer0-node-type - change to l0-node-type to make it consistent with the rest of the l0 abbreviation

haomianzheng commented:
> layer0-node-type - change to l0-node-type to make it consistent with the rest of the l0 abbreviation

This is raised during the feedback to YANG doctor, and will be revised after consensus. 
Following changes would be needed on, 
layer0-node-type -> l0-node-type
layer0-grid-type -> l0-grid-type
layer0-bandwidth-type -> l0-bandwidth-type
layer0-label-range-info -> l0-label-range-info

haomianzheng committed:
Add files via upload

Add more reference, per request in https://github.com/ietf-ccamp-wg/draft-ietf-ccamp-layer0-types/issues/4.

italobusi commented on ietf-layer0-types.yang:
```diff
@@ -43,7 +44,7 @@ module ietf-layer0-types {
      the RFC itself for full legal notices.";
 
-  revision "2019-11-28" {
+  revision "2019-12-31" {
```
```suggestion
  revision "2020-01-27" {
```

italobusi commented on ietf-layer0-types.yang:
```diff
@@ -60,106 +61,121 @@ module ietf-layer0-types {
-      (PCE) Control of Wavelength Switched Optical Networks (WSONs)";
+      (PCE) Control of Wavelength Switched Optical Networks (WSONs),
+      ITU-T G.694.1 (02/2012): Spectral grids for WDM applications: 
+      DWDM frequency grid";
```
See open issue #6

italobusi commented on ietf-layer0-types.yang:
```diff
@@ -60,106 +61,121 @@ module ietf-layer0-types {
+      of Flexi-Grid Dense Wavelength Division Multiplexing (DWDM) 
+      Networks.
+      ITU-T G.694.1 (02/2012): Spectral grids for WDM applications: 
+      DWDM frequency grid";
```
See open issue #6

italobusi commented on ietf-layer0-types.yang:
```diff
@@ -60,106 +61,121 @@ module ietf-layer0-types {
+      "RFC6163:Framework for GMPLS and Path Computation Element 
+      (PCE) Control of Wavelength Switched Optical Networks (WSONs),
+      ITU-T G.694.1 (02/2012): Spectral grids for WDM applications: 
+      DWDM frequency grid";
```
See open issue #6

italobusi commented on ietf-layer0-types.yang:
```diff
@@ -60,106 +61,121 @@ module ietf-layer0-types {
+      "RFC6163:Framework for GMPLS and Path Computation Element 
+      (PCE) Control of Wavelength Switched Optical Networks (WSONs),
+      ITU-T G.694.1 (02/2012): Spectral grids for WDM applications: 
+      DWDM frequency grid";
```
See open issue #6

italobusi commented on ietf-layer0-types.yang:
```diff
@@ -204,7 +220,9 @@ module ietf-layer0-types {
-      (PCE) Control of Wavelength Switched Optical Networks (WSONs)";
+      (PCE) Control of Wavelength Switched Optical Networks (WSONs),
+      ITU-T G.694.1 (02/2012): Spectral grids for WDM applications: 
+      DWDM frequency grid";
```
As far as I know, there is no generic definition of grid type in ITU-T but only of DWDM and CWDM grids in G.694.1 and G.694.2:

```suggestion
      DWDM frequency grid,
      ITU-T G.694.2 (12/2003): Spectral grids for WDM applications: 
      CWDM wavelength grid";
```

We can clarify this point within the draft text

haomianzheng committed:
Update ietf-layer0-types.yang

Co-Authored-By: italobusi <italo.busi@huawei.com>

haomianzheng committed:
Update ietf-layer0-types.yang

Co-Authored-By: italobusi <italo.busi@huawei.com>

haomianzheng committed:
Solving issue #6, #9 and some on the list

This update solves the open issue https://github.com/ietf-ccamp-wg/draft-ietf-ccamp-layer0-types/issues/6 and https://github.com/ietf-ccamp-wg/draft-ietf-ccamp-layer0-types/issues/9

This update also solves the  comments raised on the ccamp mailing list: 
https://mailarchive.ietf.org/arch/msg/ccamp/qp4YJCYKx-pNaajY31XEis4nFS0/

italobusi commented on ietf-layer0-types.yang:
```diff
@@ -44,30 +44,20 @@ module ietf-layer0-types {
-      f = 193100.000 GHz + N x 12.500 GHz,
+      f = 193100.000 GHz + N x channel-spacing (measured in GHz),
       where 193100.000 GHz (193.100000 THz) is the ITU-T 'anchor 
       frequency' for transmission over the C band.    ";
```
It would be good to describe also the channel-spacing:

```suggestion
        frequency' for transmission over the C band; and
      where 'channel-spacing' is defined by the dwdm-ch-spc-type.";
```

italobusi commented:
For alignment with dwdm-n and flexi-n, it would be better to introduce the channel spacing also for the cwdm-n:

https://github.com/ietf-ccamp-wg/draft-ietf-ccamp-layer0-types/blob/10acf682d70a734de21a81868f4f13a2b9c1a8f3/ietf-layer0-types.yang#L75

Proposed changes:

```
        Wavelength (nm) = 1471 + N x channel-spacing (measured in nm),
      where 1471 nm is the ITU-T 'anchor wavelength'
        for transmission over the C band; and
      where 'channel-spacing' is defined by the cwdm-ch-spc-type.";
```

italobusi commented on ietf-layer0-types.yang:
```diff
@@ -44,30 +44,20 @@ module ietf-layer0-types {
       central frequency.
       The nominal central frequency, 'f' is defined by,
-      f = 193100.000 GHz + N x 12.500 GHz,
+      f = 193100.000 GHz + N x channel-spacing (measured in GHz),
```
Editorial:

```suggestion
        f = 193100.000 GHz + N x channel-spacing (measured in GHz),
```

sergiobelotti replied:
I would prefer the word **"nominal central frequency granularity**" instead of "channel spacing" following RFC7698 definitions.

haomianzheng replied:
> I would prefer the word **"nominal central frequency granularity**" instead of "channel spacing" following RFC7698 definitions.

I think the nominal central frequency granularity is not the right term here, and insist on using "channel spacing". RFC7698 is referencing G.694.1, and I paste the following text from G.694.1: 

Section 6 for WSON: 

_- For channel spacings of 12.5 GHz on a fibre, the allowed channel frequencies (in THz) are defined by:   193.1 + n × 0.0125 where n is a positive or negative integer including 0_ 
_- For channel spacings of 25 GHz on a fibre, the allowed channel frequencies (in THz) are defined by:   193.1 + n × 0.025 where n is a positive or negative integer including 0_ 
_- For channel spacings of 50 GHz on a fibre, the allowed channel frequencies (in THz) are defined by:   193.1 + n × 0.05 where n is a positive or negative integer including 0_ 
_- For channel spacings of 100 GHz or more on a fibre, the allowed channel frequencies (in THz) are defined by:   193.1 + n × 0.1 where n is a positive or negative integer including 0_ 

Section 7 for Flexi-grid:

_- For the flexible DWDM grid, the allowed frequency slots have a nominal central frequency (in THz) defined by:   193.1 + n × 0.00625 where n is a positive or negative integer including 0  and 0.00625 is the nominal central frequency granularity in THz_ 

From the above we can see the channel spacing have a set of value and can be different according to deployment. However the term norminal central frequency granularity is only applicable to flexi-grid and have only one value as 6.25GHz. Therefore using channel spacing here is more accurate to capture what is specified in G.694.1.

italobusi replied:
I have checked G.694.1 and I can confirm that the term channel spacing is the right term defined in ITU-T for the 12.5 GHz/25 GHz/50 GHz/100 GHz fixed grids.

However, in case of flexible grid, the term nominal central frequency granularity is used, in clause 7, to describe the 6.25 GHz granularity of the flexible grid definition and in Appendix I.2 to describe the as well as any implementation constraints:

> For example, an application could be defined where the nominal central frequency granularity is 12.5 GHz (by only requiring values of n that are even) and that only requires slot widths as a multiple of 25 GHz (by only requiring values of m that are even).

A similar text is also reported in section 3.4.1 of RFC7698:

>    As stated in Clause 7 of [G.694.1], the flexible DWDM grid has a
>    nominal central frequency granularity of 6.25 GHz and a slot width
>    granularity of 12.5 GHz.  However, devices or applications that make
>    use of the flexible grid might not be capable of supporting every
>    possible slot width or position.  In other words, applications may be
>    defined where only a subset of the possible slot widths and positions
>    is required to be supported.  For example, an application could be
>    defined where the nominal central frequency granularity is 12.5 GHz
>    (by only requiring values of n that are even) and where slot widths
>    are a multiple of 25 GHz (by only requiring values of m that are
>    even).
> 

From a YANG modelling perspective we need two different attributes to represent the 6,25 GHz and the implementation constraints (e.g., the 12.5 GHz)

RFC6205 re-uses the term channel spacing (C.S.) to represent the 6.25GHz and RFC8363 defines the nominal central frequency granularity as:

>    C.F.G (Central Frequency Granularity, 8 bits): A positive integer.
>    Its value indicates the multiple of C.S., in terms of central
>    frequency granularity.
> 

Based on this analysis, I think we could use the term channel spacing to represent the 6.25GHz; redefine the nominal-central-frequency-granularity attribute as an uint8 (in alignment with RFC8363) and move this attribute within the flexi-grid-label-step grouping

@haomianzheng @sergiobelotti : what do you think?

dieterbeller replied:
Looking at the many different groupings defined for DWDM fixed and flexible grid according G.694.1, I am wondering whether this is the right approach. Flexible grid is a superset of fixed grid and even for flexible grid there can be implementation constraints for n and m values. So, a generalized grouping could be defined for both fixed and flexible grid where n and m constraints are defined and all cases are covered. This would make many different micro groupings obsolete.

sergiobelotti replied:
@italobusi  : 
not clear to me your proposal : what do you want to say with "redefine the nominal-central-frequency-granularity attribute as an uint8 (in alignment with RFC8363) and move this attribute within the flexi-grid-label-step grouping". 
Btw I think we should avoid to go on with small separate grouping between flex grid and fixed grid : we are not talking about two different things but one appear to me a superset of the other so we should try as much as possible to have a common structure to represent both and only , when there is not alternative, using specific grouping.

@italobusi @haomianzheng 
I think this discussion about the right definition with respect "granularity" in the nominal central frequency context, is a further example of the confusion , unfortunately, inside the WSON, FlexGrid model. If you take e.g. RFC 7699  it is even using internal at the same section 2 about "Overview of Flexi-Grid" the two terms mixing them
" The basis of the work is to allow a data channel to be formed
   from an abstract grid anchored at 193.1 THz and selected on a channel
   spacing of 6.25 GHz with a variable slot width measured in units of
   12.5 GHz.  Individual allocations may be made on this basis from
   anywhere in the spectrum, subject to allocations not overlapping.

   [G.694.1] provides clear guidance on the support of flexible grid by
   implementations in Section 2 of Appendix I:

      The flexible DWDM grid defined in clause 7 has a nominal central
      frequency granularity of 6.25 GHz and a slot width granularity of
      12.5 GHz.  However, devices or applications that make use of the
      flexible grid may not have to be capable of supporting every
      possible slot width or position.  In other words, applications may
      be defined where only a subset of the possible slot widths and
      positions are required to be supported.

      For example, an application could be defined where the nominal
      central frequency granularity is 12.5 GHz (by only requiring
      values of n that are even) and that only requires slot widths as a
      multiple of 25 GHz (by only requiring values of m that are even)."
and then suggesting to look at RFC 7698...
"Some additional background on the use of GMPLS for flexible grids can
   be found in [RFC7698].".
Now, frankly speaking I do not care what terms your prefer to use, obviously both are fine, my suggestion was only related to a couple of considerations:
1) trying to avoid proliferation of different definitions 
2) I consider RFC 7698 one of the best L0 document written in IETF and referencing that surely could be an assurance of avoiding misunderstanding.
The fact that "actually" nominal central frequency granularity/channel spacing in flex grid is 6.25 does not mean that it could be different in short time adding new values (as required also in IETF) , so in my view it should not associate different definitions (nsfg vs. cs) to the technology (flexible vs. fixed).

italobusi commented on ietf-layer0-types.yang:
```diff
@@ -96,7 +86,7 @@ module ietf-layer0-types {
       central frequency.
       The nominal central frequency, 'f' is defined by,
-      f = 193100.000 GHz + N x 6.250 GHz,
+      f = 193100.000 GHz + N x channel-spacing (measured in GHz),
```
Editorial:

```suggestion
        f = 193100.000 GHz + N x channel-spacing (measured in GHz),
```

italobusi commented on ietf-layer0-types.yang:
```diff
@@ -96,7 +86,7 @@ module ietf-layer0-types {
-      f = 193100.000 GHz + N x 6.250 GHz,
+      f = 193100.000 GHz + N x channel-spacing (measured in GHz),
       where 193100.000 GHz (193.100000 THz) is the ITU-T 'anchor 
       frequency' for transmission over the C band.    ";
```
It would be good to describe also the channel-spacing:

```suggestion
        frequency' for transmission over the C band; and
      where 'channel-spacing' is defined by the flexi-ch-spc-type.";
```

haomianzheng replied:
```suggestion
        frequency' for transmission over the C band; and
      where 'channel-spacing' is defined by the flexi-ch-spc-type.
      
      Note that the term 'chanel-spacing' can be alternated by the
      term 'nominal central frequency granularity' defined in
      clause 7 of ITU-T G.694.1.";
```

italobusi commented on ietf-layer0-types.yang:
```diff
@@ -112,7 +102,8 @@ module ietf-layer0-types {
        constrained to be M x SWG (that is, M x 12.500 GHz), 
-       where M is a positive integer. ";
+       where M is a positive integer. 
+	   Noted that the 12.500GHz is the 2 x channel-spacing (6.25G)";
```
Editorial:

```suggestion

       Note that the 12.500 GHz is 2 x channel-spacing (6.25 GHz)";
```

sergiobelotti replied:
Also here substitute channel spacing with "nominal central frequency granularity".

haomianzheng replied:
Better remove the note as well. There is no such notes in ITU-T G.694.1.

italobusi commented on ietf-layer0-types.yang:
```diff
@@ -121,92 +112,6 @@ module ietf-layer0-types {
-       wavelength randomly";
-  }
-
   identity least-loaded-wavelength-assignment {
```
Also this identity should be removed (its base identity has been removed)

italobusi replied:
Resolved with https://github.com/ietf-ccamp-wg/draft-ietf-ccamp-layer0-types/pull/3/commits/519087caaf7c6c554787a14c10b9b605a898c2e3

italobusi commented on ietf-layer0-types.yang:
```diff
@@ -486,59 +244,7 @@ module ietf-layer0-types {
-      "No FEC";
-  }
-
   /* Groupings. */
```
Editorial:

```suggestion
  /* Groupings. */
  
```

italobusi commented on ietf-layer0-types.yang:
```diff
@@ -814,8 +475,12 @@ module ietf-layer0-types {
-             Minimum slot width (GHz) = 
+          "Slot width range: two multipliers of the slot width ,
+           granularity, each indicating the minimal and maximal slot 
+           width supported by a port, respectively. 
+		  
+		   Minimum slot width is calculated by:
+           Minimum slot width (GHz) = 
                min-slot-width-factor * slot-width-granularity";
```
Editorial:

```suggestion
           width supported by a port, respectively. 

           Minimum slot width is calculated by:
             Minimum slot width (GHz) = 
               min-slot-width-factor * slot-width-granularity";
```

italobusi commented on ietf-layer0-types.yang:
```diff
@@ -827,8 +492,12 @@ module ietf-layer0-types {
-             Maximum slot width (GHz) = 
+          "Slot width range: two multipliers of the slot width ,
+           granularity, each indicating the minimal and maximal slot 
+           width supported by a port, respectively.
+		   
+		   Maximum slot width is calculated by:
+           Maximum slot width (GHz) = 
                max-slot-width-factor * slot-width-granularity";
```
Editorial:

```suggestion
           width supported by a port, respectively.

           Maximum slot width is calculated by:
             Maximum slot width (GHz) = 
               max-slot-width-factor * slot-width-granularity";
```

italobusi reviewed (approved):
I have just few editorial comments aimed at improving the readability of the code

haomianzheng committed:
Update ietf-layer0-types.yang

Co-Authored-By: italobusi <italo.busi@huawei.com>

haomianzheng committed:
Update ietf-layer0-types.yang

Co-Authored-By: italobusi <italo.busi@huawei.com>

haomianzheng committed:
Update ietf-layer0-types.yang

Co-Authored-By: italobusi <italo.busi@huawei.com>

haomianzheng committed:
Update ietf-layer0-types.yang

haomianzheng committed:
Update ietf-layer0-types.yang

Co-Authored-By: italobusi <italo.busi@huawei.com>

haomianzheng committed:
Update ietf-layer0-types.yang

Co-Authored-By: italobusi <italo.busi@huawei.com>

haomianzheng committed:
Update ietf-layer0-types.yang

Co-Authored-By: italobusi <italo.busi@huawei.com>

haomianzheng commented on ietf-layer0-types.yang:
```diff
@@ -43,152 +44,78 @@ module ietf-layer0-types {
-      "Wavelength selection base";
+      "M is used to determine the slot width. A slot width is 
+       constrained to be M x SWG (that is, M x 12.500 GHz), 
+       where M is a positive integer. 
+	   Noted that the 12.500GHz is the 2 x channel-spacing (6.25G)";
```
```suggestion
       where M is a positive integer.";
```

haomianzheng committed:
Update ietf-layer0-types.yang

haomianzheng commented on ietf-layer0-types.yang:
```diff
@@ -43,152 +44,77 @@ module ietf-layer0-types {
     description
       "The given value 'N' is used to compute the channel
       wavelength as per the formula:
         Wavelength (nm) = 1471 + N x 20 ";
```
```suggestion
        Wavelength (nm) = 1471 + N x channel-spacing (measured in nm),
      where 1471 nm is the ITU-T 'anchor wavelength'
        for transmission over the C band; and
      where 'channel-spacing' is defined by the cwdm-ch-spc-type. ";
```

haomianzheng committed:
Update ietf-layer0-types.yang

haomianzheng committed:
Update ietf-layer0-types.yang

haomianzheng committed:
Update ietf-layer0-types.yang

italobusi commented on ietf-layer0-types.yang:
```diff
@@ -43,308 +44,136 @@ module ietf-layer0-types {
      the RFC itself for full legal notices.";
 
-  revision "2019-11-28" {
+  revision "2020-04-15" {
```
```suggestion
  revision "2020-04-20" {
```

italobusi reviewed (approved):
Please update the revision date

haomianzheng committed:
Update ietf-layer0-types.yang

Co-Authored-By: italobusi <italo.busi@huawei.com>

haomianzheng commented:
> layer0-node-type - change to l0-node-type to make it consistent with the rest of the l0 abbreviation

This has been done in the latest commit.

sergiobelotti commented on ietf-layer0-types.yang:
```diff
@@ -597,59 +380,16 @@ module ietf-layer0-types {
-      type int16;
+      type l0-types:flexi-n;
       description
         "The central frequency in Flexi-grid.";
```
for l0-types:flexi-n the description should be the same as per the typedef of flexi-n that means :"The given value 'N' is used to determine the nominal central frequency." . This is exactly the same what happen for flexi-m leaf, below in the grouping. "M is used to determine the slot width. A slot width is constrained to be M x SWG (that is, M x 12.500 GHz), where M is a positive integer.";

sergiobelotti commented on ietf-layer0-types.yang:
```diff
@@ -756,31 +479,43 @@ module ietf-layer0-types {
+      description
+        "The restriction of flexi-n must be certain multipliers 
+          of channel-spacing. This is called the central frequency
+          granularity in RFC8363. ";
```
Could you re-phrase the description of this leaf not sure to have understood what you mean here ? It is mentioned central frequency granularity that is a definition specific for RFC 8363. This RFC says that for any "definitions" the reference IETF is RFC 7698. I do not like central frequency granularity and we need to be more clear in the description.

italobusi replied:
@sergiobelotti : what about replacing the description with this one?

```
This attribute defines the granularity of supported values for the nominal central frequency as a multiplier of the channel-spacing.

For example, given a grid with a nominal central frequency granularity of 6.25 GHz, the granularity of the supported values of the nominal central frequency could be 12.5 GHz. In this case, the values of flexi-n should be even and this constraints is reported by setting the flexi-n-step to 2.

This attribute is also known as central frequency granularity in RFC8363.
```

haomianzheng committed:
Update ietf-layer0-types.yang
